### PR TITLE
Launchpad PRO support

### PR DESCRIPTION
**Pull request by mrVanDalo**
A public fork aimed to Novation Launchpad PRO users

### CODE REVIEW

**Comment by mrVanDalo** [comment]
@The-XOR I guess you tested it with your launchpad pro, right?

I merge it in a feature branch and revert some of your changes, especially the name and logo stuff.
I also make it possible that both versions can live in the same project.
 
It looks like you are using windows. I try to combine your changes so it will run under windows and linux

I don't understand the thing with the different in and out device.

File: src/Layer/Launchpad/Launchpad.lua [hunk 1]
```diff
@@ -39,9 +39,9 @@ end
 
 --- connect launchpad to a midi device
 --
-function Launchpad:connect(midi_device_name)
-    self.midi_out    = renoise.Midi.create_output_device(midi_device_name)
-    self.midi_input  = renoise.Midi.create_input_device(midi_device_name, self.__midi_input_callback)
+function Launchpad:connect(midi_in_device_name, midi_out_device_name)
```
Comment (mrVanDalo) [inline src/Layer/Launchpad/Launchpad.lua]: what do you need the out_device here? why should the render device and the input device differ?

File: src/Layer/Launchpad/Launchpad.lua [hunk 2]
```diff
@@ -118,9 +118,30 @@ function Launchpad:send(channel, number, value)
     if (not self.midi_out or not self.midi_out.is_open) then
         return
     end
-    local message = {channel, number, value}
-    -- print(("Launchpad : send MIDI %X %X %X"):format(message[1], message[2], message[3]))
-    self.midi_out:send(message)
+	if (channel == 0x90 and value > 127) then
```
Comment (mrVanDalo) [inline src/Layer/Launchpad/Launchpad.lua]: what is this for?

**Comment by The-XOR** [comment]
that's ok, I changed the name just to have a unique fork name.
the two ports are needed because launchpad pro names are DIFFERENT for midi
in and midi out ports, so you must have two selections and open two
differnt ports.
as you can see, the layout of lp pro is completely different. for easyness,
I just inserted a filter that remaps the input as it if was an mk1 and does
the opposite on output, leaving the rest of the code untouched; the out
filter also take care of remapping colors and flash codes.

mauro



Il domenica 28 febbraio 2016, Ingolf Wagner <notifications@github.com> ha
scritto:

**Comment by mrVanDalo** [comment]
Can we automatically find out how the different devices are named? So we can automatically create the in and output port of it.

**Comment by The-XOR** [comment]
in theory yes, but if you have multiple lp pro attached ( I have two) it is
not so easy. this because EVERY lp creates three input and three output
ports, and even worse this name changed when I upgraded my launchpads.
I can send you the name scheme, but I'm not on the pc, now.

Il domenica 28 febbraio 2016, Ingolf Wagner <notifications@github.com> ha
scritto:

File: src/Layer/Launchpad/Launchpad.lua (hunk 2 above)
Comment (The-XOR) [inline src/Layer/Launchpad/Launchpad.lua]: this is for flash buttons. flash is handled with sysex in lp pro, so bit 7
is used for signaling flash, while bits 0-6 are the color code.

Il domenica 28 febbraio 2016, Ingolf Wagner <notifications@github.com> ha
scritto:

**Comment by mrVanDalo** [comment]
Ok, thanks for the information. 

You made it work with these patches right?

Anything else which is important? like special settings for the launchpad or something?

File: src/Layer/Launchpad/Launchpad.lua (hunk 2 above)
Comment (mrVanDalo) [inline src/Layer/Launchpad/Launchpad.lua]: Ok, I see! but all other messages are handled like before, just with different velocities, right?

**Comment by The-XOR** [comment]
yes, but I (still) cannot say that it was 100% tested. the only advice is
that lp pro must be set in User Mode, later on I will browse the manual and
see if there is a sysex command that can do it.
As a professional developer, I am quite LAZY when I have to code for my
favourite hobby: I feel like I was still working  :-)

regards,
mauro

Il domenica 28 febbraio 2016, Ingolf Wagner <notifications@github.com> ha
scritto:

File: src/Layer/Launchpad/Launchpad.lua (hunk 2 above)
Comment (The-XOR) [inline src/Layer/Launchpad/Launchpad.lua]: The new velocities MORE OR LESS map the same colors as the original
launchpad.
even here, novation decided to make things completely incompatible....

Il domenica 28 febbraio 2016, Ingolf Wagner <notifications@github.com> ha
scritto:

**Comment by mrVanDalo** [comment]
hehe. 

yeah, sometimes it feels like working, but most of the time I'm using it, so it pays :D

File: src/Layer/Launchpad/Launchpad.lua (hunk 2 above)
Comment (mrVanDalo) [inline src/Layer/Launchpad/Launchpad.lua]: Ok, I will create a more complex color set (according to the features of the launchpad pro) and map it to the old launchpad, this way people can enjoy the new features of the new launchpad. But not in this PR. This time I just want to make it work.

**Pull request closed without merging**